File: sdk/samples/01_basics/python/hello_world.py
#!/usr/bin/env python3
# SPDX-License-Identifier: Apache-2.0 OR MIT
# Copyright (c) 2025-2026 naskel.com

"""
HDDS Sample: Hello World (Python)

Demonstrates basic pub/sub with HDDS Python API.

Usage:
    # Terminal 1 - Subscriber
    python hello_world.py

    # Terminal 2 - Publisher
    python hello_world.py pub
"""

import sys
import time

# Add SDK to path
sys.path.insert(0, '../../../python')

import hdds
from generated.HelloWorld import HelloWorld


def run_publisher(participant):
    """Run publisher sending HelloWorld messages."""
    print("Creating writer...")
    writer = participant.create_writer("HelloWorldTopic")

    print("Publishing messages...")
    for i in range(10):
        msg = HelloWorld(message="Hello from HDDS Python!", count=i)
        data = msg.serialize()
        writer.write(data)
        print(f"  Published: {msg.message} (count={msg.count})")
        time.sleep(0.5)

    print("Done publishing.")


def run_subscriber(participant):
    """Run subscriber receiving HelloWorld messages."""
    print("Creating reader...")
    reader = participant.create_reader("HelloWorldTopic")

    # Create waitset for efficient waiting
    waitset = hdds.WaitSet()
    cond = reader.get_status_condition()
    waitset.attach(cond)

    print("Waiting for messages (Ctrl+C to exit)...")
    received = 0

    while received < 10:
        # Wait up to 5 seconds
        if waitset.wait(timeout=5.0):
            # Take all available samples
            while True:
                data = reader.take()
                if data is None:
                    break
                msg, _ = HelloWorld.deserialize(data)
                print(f"  Received: {msg.message} (count={msg.count})")
                received += 1
        else:
            print("  (timeout - no messages)")

    print("Done receiving.")


def main():
    is_publisher = len(sys.argv) > 1 and sys.argv[1].lower() in ('pub', 'publisher', '-p')

    # Initialize logging
    hdds.logging.init(hdds.LogLevel.INFO)

    # Create participant
    print("Creating participant...")
    participant = hdds.Participant("HelloWorld")

    try:
        if is_publisher:
            run_publisher(participant)
        else:
            run_subscriber(participant)
    except KeyboardInterrupt:
        print("\nInterrupted.")
    finally:
        print("Cleanup complete.")


if __name__ == "__main__":
    main()

File: sdk/samples/01_basics/c/hello_world.c
// SPDX-License-Identifier: Apache-2.0 OR MIT
// Copyright (c) 2025-2026 naskel.com

/**
 * HDDS Sample: Hello World (C)
 *
 * Demonstrates basic pub/sub with HDDS C API.
 *
 * Usage:
 *     # Terminal 1 - Subscriber
 *     ./hello_world
 *
 *     # Terminal 2 - Publisher
 *     ./hello_world pub
 */

#include <hdds.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

#include "generated/HelloWorld.h"

static void run_publisher(HddsParticipant* participant) {
    printf("Creating writer...\n");
    HddsDataWriter* writer = hdds_writer_create(participant, "HelloWorldTopic");
    if (!writer) {
        fprintf(stderr, "Failed to create writer\n");
        return;
    }

    printf("Publishing messages...\n");
    HelloWorld msg;
    HelloWorld_init(&msg);
    strcpy(msg.message, "Hello from HDDS C!");

    for (int i = 0; i < 10; i++) {
        msg.id = i;

        // Serialize and write
        uint8_t buffer[1024];
        size_t len = HelloWorld_serialize(&msg, buffer, sizeof(buffer));

        if (hdds_writer_write(writer, buffer, len) == HDDS_OK) {
            printf("  Published: %s (id=%d)\n", msg.message, msg.id);
        } else {
            fprintf(stderr, "  Failed to publish message %d\n", i);
        }

        usleep(500000);  // 500ms
    }

    printf("Done publishing.\n");
    hdds_writer_destroy(writer);
}

static void run_subscriber(HddsParticipant* participant) {
    printf("Creating reader...\n");
    HddsDataReader* reader = hdds_reader_create(participant, "HelloWorldTopic");
    if (!reader) {
        fprintf(stderr, "Failed to create reader\n");
        return;
    }

    // Create waitset
    HddsWaitSet* waitset = hdds_waitset_create();
    const HddsStatusCondition* cond = hdds_reader_get_status_condition(reader);
    hdds_waitset_attach_status_condition(waitset, cond);

    printf("Waiting for messages (Ctrl+C to exit)...\n");
    int received = 0;

    while (received < 10) {
        // Wait up to 5 seconds
        const void* triggered[1];
        size_t count;

        if (hdds_waitset_wait(waitset, 5000000000LL, triggered, 1, &count) == HDDS_OK && count > 0) {
            // Take all available samples
            uint8_t buffer[1024];
            size_t len;

            while (hdds_reader_take(reader, buffer, sizeof(buffer), &len) == HDDS_OK) {
                HelloWorld msg;
                HelloWorld_deserialize(&msg, buffer, len);
                printf("  Received: %s (id=%d)\n", msg.message, msg.id);
                received++;
            }
        } else {
            printf("  (timeout - no messages)\n");
        }
    }

    printf("Done receiving.\n");
    hdds_waitset_destroy(waitset);
    hdds_reader_destroy(reader);
}

int main(int argc, char** argv) {
    int is_publisher = (argc > 1 &&
        (strcmp(argv[1], "pub") == 0 ||
         strcmp(argv[1], "publisher") == 0 ||
         strcmp(argv[1], "-p") == 0));

    // Initialize logging
    hdds_logging_init(3);  // INFO level

    // Create participant
    printf("Creating participant...\n");
    HddsParticipant* participant = hdds_participant_create("HelloWorld");
    if (!participant) {
        fprintf(stderr, "Failed to create participant\n");
        return 1;
    }

    if (is_publisher) {
        run_publisher(participant);
    } else {
        run_subscriber(participant);
    }

    // Cleanup
    hdds_participant_destroy(participant);
    printf("Cleanup complete.\n");

    return 0;
}
